test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperPasswordTest.cs

```diff
@@ -439,6 +439,53 @@ private static ViewDataDictionary<PasswordModel> GetViewDataWithErrors()
             return viewData;
         }
 
+        public static TheoryData PasswordFor_UsesModelStateValueForComplexExpressionsData
+        {
+            get
+            {
+                return new TheoryData<Expression<Func<PasswordModel, string>>, string>
+                {
+                    {
+                        model => model.Property3["key"],
+                        @"<input id=""HtmlEncode[[pre_Property3_key_]]"" name=""HtmlEncode[[pre.Property3[key]]]"" " +
+                        @"type=""HtmlEncode[[password]]"" />"
+                    },
+                    {
+                        model => model.Property4.Property5,
+                        @"<input id=""HtmlEncode[[pre_Property4_Property5]]"" name=""HtmlEncode[[pre.Property4.Property5]]"" " +
+                        @"type=""HtmlEncode[[password]]"" />"
+                    },
+                    {
+                        model => model.Property4.Property6[0],
+                        @"<input id=""HtmlEncode[[pre_Property4_Property6_0_]]"" " +
+                        @"name=""HtmlEncode[[pre.Property4.Property6[0]]]"" type=""HtmlEncode[[password]]"" />"
+                    }
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(PasswordFor_UsesModelStateValueForComplexExpressionsData))]
+        public void PasswordFor_ComplexExpressions_IgnoresValueFromViewData(
+            Expression<Func<PasswordModel, string>> expression,
+            string expected)
+        {
+            // Arrange            
+            var model = new PasswordModel();
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(model);
+            helper.ViewData.TemplateInfo.HtmlFieldPrefix = "pre";
+            helper.ViewData.Model.Property3["key"] = "Prop3Val";
+            helper.ViewData.Model.Property4.Property5 = "Prop5Val";
+            helper.ViewData.Model.Property4.Property6.Add("Prop6Val");
```

Comment:
Name of test and name of the member data should be similar since they align 1:1. Further both names are incorrect since setup touches neither `ModelState` nor the dictionary behind `ViewData`. Test is really showing that `PasswordFor()` ignores the expression value. Should touch `ViewData["Property3"]` in addition to the current setup. Then name the test and member data appropriately i.e. after fact `PasswordFor()` ignores everything but `ModelState`.

Separately, do we already have a test that shows `PasswordFor()` getting a `value` from `ModelState`?